File: lintcode/python/0460_find_k_closest_elements.py
#!/usr/bin/python -t

# binary search solution
#采用的是二分法 + 双指针
#二分法确定一个位置，左侧是 < target，右侧是 >= target
#然后用两根指针从中间向两边走，依次找到最接近的 k 个数

class Solution:
    """
    @param A: an integer array
    @param target: An integer
    @param k: An integer
    @return: an integer array
    """
    def kClosestNumbers(self, A, target, k):
        # write your code here
        n = len(A)
        l = 0
        r = n-1
        target_index = -1
        
        while l < r:
            mid = (l+r)/2
            if A[mid] == target:
                target_index = mid
                break
            elif A[mid] < target:
                l = mid+1
            else:
                r = mid
        if target_index == -1:
            target_index = l
        ret = []
        l = target_index-1
        r = target_index
        
        for _ in range(k):
            if self.is_left_closer(A, target, l, r):
                ret.append(A[l])
                l -= 1
            else:
                ret.append(A[r])
                r += 1
                
        return ret
        
    def is_left_closer(self, A, target, l, r):
        if l < 0:
            return False
        if r >= len(A):
            return True
        return target-A[l] <= A[r] - target

File: lintcode/golang/0460_find_k_closest_elements.go
/**
 * @param A: an integer array
 * @param target: An integer
 * @param k: An integer
 * @return: an integer array
 */
func kClosestNumbers (A []int, target int, k int) []int {
    // write your code here
    var ret []int
    ret = []int{}
    n := len(A)
    if n == 0 {
        return ret
    }
    
    if k == 0 {
        return ret
    }
    
    l := 0
    r := n-1
    
    for l + 1 < r {
        mid := (l+r)/2
        if A[mid] >= target {
            r = mid
        } else {
            l = mid
        }
    }
    
    for k > 0 {
        if l >= 0 && r < n {
            if abs(A[l] - target) <= abs(A[r] - target) {
                ret = append(ret, A[l])
                l -= 1
            } else {
                ret = append(ret, A[r])
                r += 1
            }
        } else if l < 0 {
            ret = append(ret, A[r])
            r += 1
        } else {
            ret = append(ret, A[l])
            l -= 1
        }
        k -= 1
    }
    
    return ret
}

func abs (a int) int {
    if a < 0 {
        return -a
    }
    
    return a
}
